File: modules/aws_scanner/ec2_scanner.py
import boto3
import json
import uuid
import time
import logging
import os
from datetime import datetime
from botocore.exceptions import ClientError

# Set up logger
logger = logging.getLogger("EC2Scanner")
logger.setLevel(logging.INFO)

# Ensure log directories exist inside container
import os
os.makedirs("/logs/merope", exist_ok=True)
os.makedirs("/logs/aws_scans/ec2", exist_ok=True)

# Add file handler for logging
file_handler = logging.FileHandler("/logs/merope/scanner.log")
file_formatter = logging.Formatter('%(asctime)s [%(levelname)s] %(message)s')
file_handler.setFormatter(file_formatter)
logger.addHandler(file_handler)

# Add console handler
console_handler = logging.StreamHandler()
console_handler.setFormatter(file_formatter)
logger.addHandler(console_handler)

# Custom JSON Encoder to handle datetime objects
class CustomJSONEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, datetime):
            return obj.isoformat()  # Convert datetime to ISO string
        elif isinstance(obj, bytes):
            return obj.decode('utf-8', errors='ignore')  # Handle binary output
        return super().default(obj)

# Generate unique scan ID
def generate_scan_id():
    return f"SCAN-{uuid.uuid4()}-{int(time.time())}"

# Discover active AWS regions
def get_active_regions(session):
    try:
        ec2_client = session.client('ec2', region_name='us-east-1')
        response = ec2_client.describe_regions()
        return [region['RegionName'] for region in response['Regions']]
    except Exception as e:
        logger.warning(f" Could not fetch active regions: {e}")
        return ['us-east-1']

# Check if instance is orphaned (missing tags)
def is_orphaned(instance):
    return len(instance.get('Tags', [])) == 0

# Get associated resources (Security Groups, Volumes, ENIs)
def get_associated_resources(ec2, instance):
    result = {
        "security_groups": [],
        "volumes": [],
        "network_interfaces": []
    }

    sg_ids = [sg['GroupId'] for sg in instance.get('SecurityGroups', [])]
    if sg_ids:
        try:
            result['security_groups'] = ec2.describe_security_groups(GroupIds=sg_ids).get('SecurityGroups', [])
        except Exception as e:
            logger.error(f" Error fetching security groups: {e}")

    try:
        result['volumes'] = ec2.describe_volumes(Filters=[{
            'Name': 'attachment.instance-id',
            'Values': [instance['InstanceId']]
        }]).get('Volumes', [])
    except Exception as e:
        logger.error(f" Error fetching volumes: {e}")

    try:
        result['network_interfaces'] = ec2.describe_network_interfaces(Filters=[{
            'Name': 'attachment.instance-id',
            'Values': [instance['InstanceId']]
        }]).get('NetworkInterfaces', [])
    except Exception as e:
        logger.error(f" Error fetching ENIs: {e}")

    return result

# Send message to RabbitMQ
def send_to_rabbitmq(message):
    try:
        import pika

        connection = pika.BlockingConnection(pika.ConnectionParameters(host='merope-rabbit'))
        channel = connection.channel()
        channel.queue_declare(queue='scan_queue', durable=True)

        channel.basic_publish(
            exchange='',
            routing_key='scan_queue',
            body=json.dumps(message, cls=CustomJSONEncoder),
            properties=pika.BasicProperties(delivery_mode=2)  # Persistent message
        )
        connection.close()
        logger.info(f"[x] Sent EC2 Instance: {message['resource_id']}")
    except Exception as e:
        logger.error(f" Failed to send to RabbitMQ: {e}")

# Scan EC2 instances in one region
def scan_ec2_instances(session, scan_id, region):
    logger.info(f" Scanning EC2 in {region}")
    results = []

    try:
        ec2 = session.client('ec2', region_name=region)

        paginator = ec2.get_paginator('describe_instances')
        page_iterator = paginator.paginate()

        for page in page_iterator:
            reservations = page.get('Reservations', [])
            logger.info(f" Found {len(reservations)} instances in {region}")

            for reservation in reservations:
                for instance in reservation.get('Instances', []):
                    if instance['State']['Name'] != 'running':
                        continue

                    item = {
                        "scan_id": scan_id,
                        "region": region,
                        "service": "ec2",
                        "resource_type": "instance",
                        "resource_id": instance['InstanceId'],
                        "data": instance,
                        "orphaned": is_orphaned(instance),
                        "public_ip": 'PublicIpAddress' in instance,
                        "timestamp": time.time(),
                        "missing_metadata": {
                            "tags_missing": len(instance.get('Tags', [])) == 0
                        },
                        "associated_resources": get_associated_resources(ec2, instance)
                    }

                    # Save raw scan to file
                    with open(f"/logs/aws_scans/ec2/{instance['InstanceId']}.json", 'w') as f:
                        json.dump(item, f, indent=2, cls=CustomJSONEncoder)

                    # Send to RabbitMQ
                    send_to_rabbitmq(item)
                    results.append(item)

        return results

    except ClientError as ce:
        logger.warning(f" EC2 not supported in {region}: {ce}")
        return []

# Main entry point
if __name__ == "__main__":
    logger.info(" Initializing AWS Session...")
    session = boto3.Session()
    scan_id = generate_scan_id()

    regions = get_active_regions(session)
    logger.info(f" Active Regions: {regions}")

    all_results = []

    for region in regions:
        results = scan_ec2_instances(session, scan_id, region)
        all_results.extend(results)

    logger.info(f" Sent {len(all_results)} running EC2 instances.")

File: modules/aws_scanner/ec2_scanner.go
package main

import (
    "context"
    "encoding/json"
    "fmt"
    "log"
    "os"
    "time"

    "github.com/aws/aws-sdk-go-v2/config"
    "github.com/aws/aws-sdk-go-v2/service/ec2"
    "github.com/confluentinc/confluent-kafka-go/v2/kafka"
)

// Instance represents the full data structure sent to Kafka
type Instance struct {
    ScanID             string                 `json:"scan_id"`
    Region             string                 `json:"region"`
    Service            string                 `json:"service"`
    ResourceType       string                 `json:"resource_type"`
    ResourceID           string                 `json:"resource_id"`
    Data               map[string]interface{}   `json:"data"`
    Orphaned           bool                   `json:"orphaned"`
    PublicIP           bool                   `json:"public_ip"`
    MissingMetadata    map[string]interface{}   `json:"missing_metadata"`
    AssociatedResources map[string]interface{} `json:"associated_resources"`
    Timestamp          int64                  `json:"timestamp"`
}

// Set up logging
var logger *log.Logger

func init() {
    logFile, err := os.OpenFile("/logs/merope/scanner.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
    if err != nil {
        log.Fatalf(" Failed to open log file: %v", err)
    }
    logger = log.New(logFile, "", log.LstdFlags)
    logger.SetPrefix("[EC2Scanner] ")
    logger.SetFlags(0)
}

func getEC2Client(region string) (*ec2.Client, error) {
    cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(region))
    if err != nil {
        return nil, fmt.Errorf("failed to load AWS config: %v", err)
    }
    return ec2.NewFromConfig(cfg), nil
}

func isOrphaned(instance map[string]interface{}) bool {
    tags, ok := instance["Tags"].([]interface{})
    return !ok || len(tags) == 0
}

func hasPublicIP(instance map[string]interface{}) bool {
    ip, ok := instance["PublicIpAddress"].(string)
    return ok && ip != ""
}

func convertInstance(instance ec2.Instance) map[string]interface{} {
    m := make(map[string]interface{})
    m["InstanceId"] = instance.InstanceId
    m["State"] = instance.State
    m["PublicIpAddress"] = instance.PublicIpAddress
    m["Tags"] = instance.Tags
    m["SecurityGroups"] = instance.SecurityGroups
    m["VpcId"] = instance.VpcId
    m["SubnetId"] = instance.SubnetId
    m["AccountId"] = instance.AccountId
    m["KeyName"] = instance.KeyName
    m["LaunchTime"] = instance.LaunchTime
    m["PlatformDetails"] = instance.PlatformDetails
    return m
}

func scanEC2Instances(scanID string) ([]Instance, error) {
    regions, err := getActiveRegions()
    if err != nil {
        return nil, err
    }

    var allInstances []Instance

    for _, region := range regions {
        client, err := getEC2Client(region)
        if err != nil {
            log.Printf(" Skipping region %s: %v", region, err)
            continue
        }

        input := &ec2.DescribeInstancesInput{}
        paginator := ec2.NewDescribeInstancesPaginator(client, input)

        for paginator.HasMorePages() {
            page, err := paginator.NextPage(context.TODO())
            if err != nil {
                log.Printf(" Error fetching page in %s: %v", region, err)
                continue
            }

            for _, reservation := range page.Reservations {
                for _, instance := range reservation.Instances {
                    instanceMap := convertInstance(*instance)
                    orphaned := isOrphaned(instanceMap)

                    item := Instance{
                        ScanID:       scanID,
                        Region:       region,
                        Service:      "ec2",
                        ResourceType: "instance",
                        ResourceID:   *instance.InstanceId,
                        Data:         instanceMap,
                        Orphaned:     orphaned,
                        PublicIP:     hasPublicIP(instanceMap),
                        MissingMetadata: map[string]interface{}{
                            "tags_missing": orphaned,
                        },
                        AssociatedResources: map[string]interface{}{
                            "security_groups": instanceMap["SecurityGroups"],
                            "vpc_id":        instanceMap["VpcId"],
                        },
                        Timestamp: time.Now().Unix(),
                    }

                    // Save raw scan to file
                    saveInstanceToFile(item)

                    // Add to results
                    allInstances = append(allInstances, item)
                }
            }
        }
    }

    return allInstances, nil
}

func saveInstanceToFile(instance Instance) {
    data, _ := json.MarshalIndent(instance, "", "  ")
    err := os.WriteFile(fmt.Sprintf("/logs/aws_scans/ec2/%s.json", instance.ResourceID), data, 0644)
    if err != nil {
        log.Printf(" Failed to save instance %s to file: %v", instance.ResourceID, err)
    }
}

func sendToKafka(topic string, instances []Instance) {
    p, err := kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": "merope-kafka:9092"})
    if err != nil {
        log.Fatalf(" Failed to create Kafka producer: %v", err)
    }
    defer p.Close()

    for _, msg := range instances {
        jsonData, _ := json.Marshal(msg)

        deliveryChan := make(chan kafka.Event)
        err := p.Produce(&kafka.Message{
            TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
            Key:            []byte(msg.ResourceID),
            Value:          jsonData,
        }, deliveryChan)

        if err != nil {
            log.Printf(" Failed to send to Kafka: %v", err)
        } else {
            e := <-deliveryChan
            m := e.(*kafka.Message)
            if m.TopicPartition.Error != nil {
                log.Printf(" Kafka delivery failed: %v", m.TopicPartition.Error)
            } else {
                log.Printf(" Sent to Kafka: %s [%d] at offset %v\n", *m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset)
            }
        }
        close(deliveryChan)
    }
}

func getActiveRegions() ([]string, error) {
    cfg, _ := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
    ec2Client := ec2.NewFromConfig(cfg)

    input := &ec2.DescribeRegionsInput{}
    output, err := ec2Client.DescribeRegions(context.TODO(), input)
    if err != nil {
        return []string{"us-east-1"}, nil
    }

    var regions []string
    for _, r := range output.Regions {
        regions = append(regions, *r.RegionName)
    }

    return regions, nil
}

func main() {
    scanID := fmt.Sprintf("SCAN-%d", time.Now().UnixNano())

    log.Println(" Initializing AWS Session...")
    log.Printf(" Scan ID: %s\n", scanID)

    instances, err := scanEC2Instances(scanID)
    if err != nil {
        log.Printf(" No running instances found")
        return
    }

    log.Printf(" Found %d running EC2 instances\n", len(instances))

    // Send to Kafka
    sendToKafka("aws_scan_data", instances)
}